login/login.go

```diff
@@ -0,0 +1,96 @@
+package login
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"github.com/juju/cmd"
+	"github.com/juju/usso"
+	"golang.org/x/crypto/ssh/terminal"
+	"gopkg.in/errgo.v1"
+)
+
+type TokenGetter interface {
+	GetTokenWithOTP(email, password, otp, tokenName string) (*usso.SSOData, error)
+}
+
+var ussoServer (TokenGetter) = usso.ProductionUbuntuSSOServer
+
+// LoginUSSO reads login parameters from ctxt.Stdin and then retrieves an
+// oauth token from Ubuntu SSO.
+func LoginUSSO(ctx *cmd.Context, twoFactor bool, tokenPath string) (*usso.SSOData, error) {
+	email, pass, otp, err := readUSSOParams(ctx, twoFactor)
+	if err != nil {
+		return nil, errgo.Notef(err, "cannot read login parameters")
+	}
+	tok, err := ussoServer.GetTokenWithOTP(email, pass, otp, "charm")
+	if err != nil {
+		return nil, errgo.Notef(err, "cannot get token")
+	}
+	if err := SaveToken(tokenPath, tok); err != nil {
+		return nil, errgo.Notef(err, "cannot save token")
+	}
+	return tok, nil
+}
+
+func readPassword(ctx *cmd.Context, br *bufio.Reader) (string, error) {
```

Comment:
Ideally this would use juju/utils ReadPassword. it might not be compatible though.